### ISSUE TITLE
[cssom] Serialization of large numbers should use scientific notation

### ISSUE THREAD

**Issue by achristensen07** [w3c/csswg-drafts#6471]
https://drafts.csswg.org/cssom/#serialize-a-css-component-value in the <number> section has a note that says "scientific notation is not used."  While this makes sense for small values so we serialize "0" instead of "6.123233995736766e-17" and this matches the behavior of Chrome and Firefox and I'm looking into implementing this in WebKit, for large numbers all browsers do use scientific notation, as illustrated by this simple example:
```
<body onload='checkLargeValues()'>
<script>
function checkLargeValues() {
    let target = document.getElementById('p');
    target.style['line-height'] = 1e+26;
    alert(window.getComputedStyle(target)['line-height'])
}
</script>
<p id="p"></p>
</body>
```
The phrase "in the shortest form possible" is already in the text, which suggests to me that multiple forms are possible.  I suggest the note be clarified to say something like "scientific notation is not used with negative exponents."

**Comment by achristensen07** [comment]
BTW I found this while trying to implement it at https://bugs.webkit.org/show_bug.cgi?id=228190 based on some earlier work by Tyler Wilcock.

**Comment by tabatkins** [comment]
Agreed, the spec should reflect reality there.

A small wrinkle is that, per [the current Syntax spec](https://drafts.csswg.org/css-syntax/#consume-a-number), using scinot automatically makes it a `<number>`, not an `<integer>`. I did this for simplicity, but if we're serializing large numbers, either properties like `z-index` need to *not* use scinot even for large numbers, or I need to amend Syntax to allow scinot to produce integers at least some of time.

(Happy to go with either "only when there's no decimal point", so `23e8` is an integer but `23.4e8` isn't, or the more forgiving "as long as the exponent exceeds the number of digits after the decimal point", so `23.4e8` would be an integer but `2.34e1` isn't.)

**Comment by dlibby-** [comment]
In terms of expected behaviors for developers, I think the current behavior for large numbers is somewhat unintuitive, as values don't roundtrip and script cannot disambiguate between two large numbers that are close to each other.

```
let value = 1000000;
document.body.style.lineHeight = value + "px";
document.documentElement.style.lineHeight = (++value) + "px";

// both '1e+06px', evals to true
document.documentElement.style.lineHeight === document.body.style.lineHeight 
```

Maybe we should specify that for numbers < X, use a set number of precision digits (in Blink I believe this is 6 currently), but numbers larger should use N precision digits where N is the number of digits (maybe +2?). Or perhaps inertia is enough that specifying the current behavior is the right call for web compat.

**Comment by cdoublev** [comment]
> I need to amend Syntax to allow scinot to produce integers at least some of time.

Copy-paste of what I was suggesting in #7289 :

> Otherwise, step 5.3 of [consume a number](https://drafts.csswg.org/css-syntax-3/#consume-a-number) could be replaced after step 6 by evaluating if `repr` is a number. I do not know how this evaluation can be translated into spec/WebIDL terms.
>
> Perhaps it can not be fixed because it would break parsing eg. [`<signless-integer>` and `<signed-integer>`](https://drafts.csswg.org/css-syntax-3/#typedef-signless-integer).

**Comment by cdoublev** [comment]
I'm not sure I totally understand the intent of the resolution for the new definition. The current definition to serialize `<number>` is:

> A base-ten number using digits 0-9 (U+0030 to U+0039) in the shortest form possible, using "." to separate decimals (if any), rounding the value if necessary to not produce more than 6 decimals, preceded by "-" (U+002D) if it is negative.

Reading the IRC log and observing values produced in Chrome and Firefox, I think the intent is to limit to 6 significant digits, not to 6 decimals. Is it correct?

Does it refer to `Number.toPrecision()` as the *serialization of JS* to match? `Number.toPrecision()` does not always produce *the shortest form possible*, eg. `(100).toPrecision(2)` is `1.0e+2` instead of `1e2`, and I also read: *Leading 0s are dropped*. Leading 0s are not dropped by `Number.toPrecision()`. 

What is the purpose of serialization in the shortest possible form? Is it to optimize the internal storage of values?

Lastly, this resolution only applies to `<number>` and not to `<integer>`, isn't it? Then this issue does not justify modifying Syntax to produce a `<number-token>` with a `type` equal to `integer` in some cases where scientific notification is used, isn't it?

**Comment by tabatkins** [comment]
> I think the intent is to limit to 6 significant digits, not to 6 decimals. Is it correct?

No, it means six digits after the decimal point. An integer in the millions shouldn't start rounding off its lower integer digits.

> Does it refer to Number.toPrecision() as the serialization of JS to match?

No, we were referring to the normal JS serialization of numbers, aka what happens when you stringify, as in `num + ""`.

> What is the purpose of serialization in the shortest possible form? Is it to optimize the internal storage of values?

No, serialization has nothing to do with internal storage, it's just what people expect out of serialization by default. When they write `1` they don't expect it to come back out as `1.000e0`.

> Lastly, this resolution only applies to <number> and not to <integer>, isn't it? Then this issue does not justify modifying Syntax to produce a <number-token> with a type equal to integer in some cases where scientific notification is used, isn't it?

The resolution doesn't pertain to that question at all; it's a separate topic. *If* the serialization I end up defining *does* use scinot for large integers, tho, then it still requires me to make some scinot strings parse to integers.

**Comment by cdoublev** [comment]
I guess limiting to 6 significant digits is ok since no more than 6 decimal places are produced, except in the last case below where *an integer in the millions shouldn't start rounding off its lower integer digits*, which is an actual browser bug.

| line-height | style Chrome | style Firefox | +n.toFixed(6) | +n.toPrecision(6) | 
| ----------- | ------------ | ------------- | ------------- | ----------------- |
| 1e-6        | 1e-06        | 0.000001      | 0.000001      | 0.000001          |
| 1e-7        | 1e-07        | 1e-7          | 0             | 1e-7              |
| 0.123456    | 0.123456     | 0.123456      | 0.123456      | 0.123456          |
| 0.1234567   | 0.123457     | 0.123457      | 0.123457      | 0.123457          |
| 123456.7    | 123457       | 123457        | 123456.7      | 123457            |
| 999999      | 999999       | 999999        | 999999        | 999999            |
| 1e6         | 1e+06        | 1000000       | 1000000       | 1000000           |
| 1234567     | 1.23457e+06  | 1234570       | 1234567       | 1234570           |
